block-layer/codecs/DAG-JSON.md

```diff
@@ -6,6 +6,17 @@ DAG-JSON supports the full [IPLD Data Model](../data-model-layer/data-model.md).
 
 ## Format
 
+### Serialization
+
+While it is unlikely de-serializers will enforce the following rules, codec implementors
```

Comment:
If this is a SHOULD requirement, de-serializers MUST NOT enforce this rule.

Reply:
What about removing this part? Why not talking only about Serializers in this section? So it would be "Codec implementors **SHOULD**…".

Reply:
I think we should phrase it as “de-serializers SHOULD preserve the encoded ordering, encoders MUST ensure keys are ordered using ....”

Specifying any sort ordering that doesn’t conform to the encoded ordering is problematic if you want to support streaming decoders.

Any MUST regarding map ordering during de-serialization is problematic across languages at this layer, because native maps in different languages do different things to the sort ordering and we don’t want to prohibit native maps from being used in IPLD. In Selectors we can use MUST language to describe the ordering because we don’t necessarily have to use native maps when de-serializing and deterministic order is a requirement for Selectors.

Taken as a whole, these constraints leave the only good option to require deterministic ordering at encode time and attempt to preserve it during decode.

Reply:
> I think we should phrase it as “de-serializers SHOULD preserve the encoded ordering, encoders MUST ensure keys are ordered using ....”

yes, but, I don't know if it's useful to get into the layer above the codec here, those are weeds that maybe we can't afford to bring into this spec. It's the second part of that statement that I think should go in. Discussing what map-ordering occurs once you're out of the serialized form is a question for the layers above this one and we should deal with that there rather than discussing it in all of these codec docs.